### PR TITLE
Make the block header struct's internals private

### PR DESCRIPTION
**Pull request by nipunn1313**
Currently, this involves a lot of explicit cloning, but we
could migrate the return types of the get_* functions to
be copies rather than references since they are mostly copy
types anyway.

I opted to eliminate the constructor in favor of using
Default::default() plus calling a bunch of setters. This
is similar to the model that a Google Protobuf client uses
and I think it looks fine.

This begins work on #1267

### CODE REVIEW

**Comment by nipunn1313** [comment]
Bam Pull Request number 2000!

File: ethcore/src/ethereum/ethash.rs [hunk 1]
```diff
@@ -171,56 +173,56 @@ impl Engine for Ethash {
...
-		try!(UntrustedRlp::new(&header.seal[0]).as_val::<H256>());
-		try!(UntrustedRlp::new(&header.seal[1]).as_val::<H64>());
+		try!(UntrustedRlp::new(&header.seal()[0]).as_val::<H256>());
+		try!(UntrustedRlp::new(&header.seal()[1]).as_val::<H64>());
 
 		// TODO: consider removing these lines.
 		let min_difficulty = self.ethash_params.minimum_difficulty;
-		if header.difficulty < min_difficulty {
-			return Err(From::from(BlockError::DifficultyOutOfBounds(OutOfBounds { min: Some(min_difficulty), max: None, found: header.difficulty })))
+		if header.difficulty().clone() < min_difficulty {
```
Comment (rphmeier) [inline ethcore/src/ethereum/ethash.rs]: should be able to just do `if header.difficulty() < &min_difficulty`

File: ethcore/src/ethereum/ethash.rs [hunk 2]
```diff
@@ -171,56 +173,56 @@ impl Engine for Ethash {
...
 		}
 
 		let difficulty = Ethash::boundary_to_difficulty(&Ethash::from_ethash(quick_get_difficulty(
 			&Ethash::to_ethash(header.bare_hash()),
 			header.nonce().low_u64(),
 			&Ethash::to_ethash(header.mix_hash())
 		)));
-		if difficulty < header.difficulty {
-			return Err(From::from(BlockError::InvalidProofOfWork(OutOfBounds { min: Some(header.difficulty), max: None, found: difficulty })));
+		if difficulty < header.difficulty().clone() {
```
Comment (rphmeier) [inline ethcore/src/ethereum/ethash.rs]: same here

File: ethcore/src/ethereum/ethash.rs [hunk 3]
```diff
@@ -171,56 +173,56 @@ impl Engine for Ethash {
...
-			header.extra_data[..] != b"dao-hard-fork"[..] {
+		if header.number() >= self.ethash_params.dao_hardfork_transition &&
+			header.number() <= self.ethash_params.dao_hardfork_transition + 9 &&
+			header.extra_data()[..] != b"dao-hard-fork"[..] {
 			return Err(From::from(BlockError::ExtraDataOutOfBounds(OutOfBounds { min: None, max: None, found: 0 })));
 		}
 
-		if header.gas_limit > 0x7fffffffffffffffu64.into() {
-			return Err(From::from(BlockError::InvalidGasLimit(OutOfBounds { min: None, max: Some(0x7fffffffffffffffu64.into()), found: header.gas_limit })));
+		if header.gas_limit().clone() > 0x7fffffffffffffffu64.into() {
```
Comment (rphmeier) [inline ethcore/src/ethereum/ethash.rs]: also here

**Comment by rphmeier** [comment]
a lot of the `clone`s in comparisons can be avoided but LLVM will probably optimize out extraneous copying. `H256` probably shouldn't be `Copy` as it isn't _that_ small imo.

**Comment by nipunn1313** [comment]
Great point. I went through and found places where we could eliminate clone() by comparing references. Unfortunately, it seems like you can't compare a &H256 with an H256 directly because of https://github.com/rust-lang/rfcs/issues/1332, but by making both sides references, it compiles nicely as @rphmeier suggested.

I also went through and fixed up the tests and tried to minimize explicit cloning. Any time + or * is used for math, I had to do some cloning. I think the llvm compiler will do a good job of optimizing those situations.

**Pull request merged**